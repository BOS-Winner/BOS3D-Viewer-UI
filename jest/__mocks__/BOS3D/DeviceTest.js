const DeviceTest = {
  isTouchDevice: () => false,
};

export default DeviceTest;
