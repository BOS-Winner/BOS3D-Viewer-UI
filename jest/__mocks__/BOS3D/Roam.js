class Roam {
  constructor(props) {
    this.props = props;
  }

  getId() {
    return this.props.id;
  }

  export() {
    return this.props;
  }
}

export default Roam;
