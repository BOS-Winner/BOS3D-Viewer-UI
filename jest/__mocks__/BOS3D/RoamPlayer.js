/* eslint-disable lines-between-class-members */
class RoamPlayer {
  constructor(props) {
    this.roam = props.roamData;
  }

  addStopPlayCallback() {}
  addKeyFrameCallback() {}
  start() {}
  stop() {}
  play() {}
  continue() {}
}

export default RoamPlayer;
