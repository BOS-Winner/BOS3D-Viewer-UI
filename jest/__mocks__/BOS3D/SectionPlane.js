/* eslint-disable lines-between-class-members */
class SectionPlane {
  constructor() {
    this.coordinateSystem = {
      hide: () => {},
    };
  }

  setDirection() {}
  setPlane() {}
  hidePlane() {}
  showPlane() {}
  exit() {}
}

export default SectionPlane;
