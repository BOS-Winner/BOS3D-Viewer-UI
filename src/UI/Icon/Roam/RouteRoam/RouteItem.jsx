import React from 'react';
import PropTypes from 'prop-types';
import NumberControler from "Base/Number";
import Ranger from "Base/Ranger";
import downloadFile from "UIutils/download";
import _ from "lodash-es";
// import Rename from './rename';
// import Modal from '../../../Base/Modal';
// import ViewManager from './ViewManager.tsx';
// import confirm from '../../../Base/confirm';
import style from "./style.less";
import { AntdIcon } from '../../../utils/utils';

const DISABLED_COLOR = '#ccc';
class RouteItem extends React.Component {
  constructor(props) {
    super(props);
    this.nameRef = React.createRef();
    this.state = {
      time: 0,
      perspIndex: 0,
      reNameVisible: false,
    };
  }

  componentDidMount() {
    this.props.roamPlayer.addKeyFrameCallback(index => {
      this.setState({
        perspIndex: index,
        time: index * this.props.recTimeLen / (this.props.routeLen - 1)
      });
    });
    this.props.roamPlayer.addStopPlayCallback(() => {
      this.props.onPlay('stop');
      this.setState({
        time: 0,
        perspIndex: 0,
      });
    });
  }

  shouldComponentUpdate(nextProps, nextState) {
    if (nextProps.recTimeLen !== this.props.recTimeLen) {
      this.setState({
        time: nextState.perspIndex * nextProps.recTimeLen / (nextProps.routeLen - 1)
      });
    }
    if (!this.props.roamPlayer.keyFrameCallback.length) {
      this.props.roamPlayer.addKeyFrameCallback(index => {
        this.setState({
          perspIndex: index,
          time: index * this.props.recTimeLen / (this.props.routeLen - 1)
        });
      });
    }
    if (!this.props.roamPlayer.stopPlayCallback.length) {
      this.props.roamPlayer.addStopPlayCallback(() => {
        this.props.onPlay('stop');
        this.setState({
          time: 0,
          perspIndex: 0,
        });
      });
    }
    return true;
  }

  onChangeTime(time) {
    this.props.onChangeFrame(Math.round(time * (this.props.routeLen - 1) / this.props.recTimeLen));
    this.setState({
      time,
    });
  }

  // rename
  handleReName = () => {
    const { reNameVisible } = this.state;
    this.setState({
      reNameVisible: !reNameVisible,
    });
  }

  render() {
    const {
      time,
      // tempRomaTimeLen,
    } = this.state;
    const {
      playingState,
      name,
      index,
      isEditing, // editing route
      // isSaveRoute, // whether to save the route（function）
      editRoute, // edit route (function)
      selectedId,
      route
    } = this.props;

    // eslint-disable-next-line compat/compat
    // 解决时间格式化后现实多位小数，精度不准确问题。
    const timefmt = `${_.floor(Math.trunc(time / 3600), 0).toString().padStart(2, '0')}:${_.floor(Math.trunc(time % 3600 / 60), 0).toString().padStart(2, '0')}:${_.floor((time % 60), 0).toString().padStart(2, '0')}`;

    return (
      <div
        className={[isEditing !== index ? `${style.routeItem}` : `${style.routeItem} ${style.routeItemActive}`, `${selectedId === index ? style.routeItemActive : ''}`].join(' ')}
      >
        <div className={style.row1}>
          <div
            role="presentation"
            title={name}
            onClick={event => {
              event.stopPropagation();
            }}
          >
            <p title={name} className={style.nameText}>{name}</p>
          </div>
          <div>
            <AntdIcon
              type="iconicon_edit"
              title="编辑"
              className={isEditing === index ? style.antdIconActive : style.antdIcon}
              onClick={() => editRoute(index)}
            />
            <AntdIcon
              type="iconicon_export"
              title="导出"
              className={isEditing === index ? style.antdIconForbiden : style.antdIcon}
              style={{ color: isEditing === index ? DISABLED_COLOR : '' }}
              onClick={ev => {
                ev.preventDefault();
                ev.stopPropagation();
                if (isEditing !== index) {
                  const { fileName, blob } = route.exportRoute();
                  downloadFile(blob, fileName);
                }
              }}
            />
            <AntdIcon
              type="icondelete"
              title="删除"
              className={isEditing === index ? style.antdIconForbiden : style.antdIcon}
              style={{ color: isEditing === index ? DISABLED_COLOR : '' }}
              onClick={ev => {
                ev.preventDefault();
                ev.stopPropagation();
                if (isEditing !== index) {
                  this.props.onDelete();
                }
              }}
            />
          </div>
        </div>
        <div className={style.row2}>
          <AntdIcon
            type={playingState === "play" ? "iconsuspend" : "iconplay"}
            title="播放"
            className={isEditing ? style.antdIconForbiden : style.antdIcon}
            style={{ color: isEditing ? DISABLED_COLOR : '' }}
            onClick={ev => {
              ev.preventDefault();
              ev.stopPropagation();
              if (isEditing === "") { // if there is no path being edited
                this.props.onPlay(playingState === "play" ? "pause" : "play");
              }
            }}
          />
          <span>{timefmt}</span>
          <Ranger
            min={0}
            max={route.roamTime}
            step={route.roamTime / (this.props.routeLen - 1)}
            value={time}
            onChange={v => { this.onChangeTime(v) }}
          />
          <NumberControler
            initNum={1}
            step={0.1}
            min={0.1}
            max={10}
            onChange={speed => {
              this.props.onChangeSpeed(speed);
            }}
          />
        </div>
      </div>
    );
  }
}

RouteItem.propTypes = {
  name: PropTypes.string.isRequired,
  onDelete: PropTypes.func.isRequired,
  onPlay: PropTypes.func.isRequired,
  onChangeSpeed: PropTypes.func.isRequired,
  onChangeFrame: PropTypes.func.isRequired,
  playingState: PropTypes.oneOf(['play', 'stop', 'pause']).isRequired,
  recTimeLen: PropTypes.number.isRequired,
  routeLen: PropTypes.number.isRequired,
  roamPlayer: PropTypes.object.isRequired,
  index: PropTypes.string.isRequired,
  editRoute: PropTypes.func.isRequired,
  isEditing: PropTypes.string.isRequired,
  selectedId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  route: PropTypes.object.isRequired,
};

export default RouteItem;
