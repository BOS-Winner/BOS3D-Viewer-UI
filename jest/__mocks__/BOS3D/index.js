import Viewer from './Viewer';
import DeviceTest from './DeviceTest';
import SectionPlane from './SectionPlane';
import Roam from './Roam';
import RoamPlayer from './RoamPlayer';

export default {
  Viewer,
  EVENTS: {
    ON_CLICK_PICK: 'ON_CLICK_PICK',
    ON_RECTPICK_ADD: 'ON_RECTPICK_ADD',
    ON_LOAD_COMPLETE: 'ON_LOAD_COMPLETE',
    ON_SELECTION_CHANGED: 'ON_SELECTION_CHANGED',
    ON_POINTERLOCK_EXIST: 'ON_POINTERLOCK_EXIST',
  },
  DeviceTest,
  SectionPlane,
  Roam,
  RoamPlayer,
  RotatePivotMode: {
    MOUSEPOINT: 0, SELECTION: 1, CENTER: 2, CAMERA: 3
  },
  ControlConfig: {
    NoRotate: false,
  },
  GlobalData: {
    EnableSelectionOutline: true,
    EnableSelectionBoundingBox: false,
    EnableSelectionByTranslucent: false,
    WalkingWithViewLock: false,
  },
  DrawingStyle: {
    SHADING: 0,
    BOARDLINE: 1,
    SHADINGWITHLINE: 2,
  }
};
