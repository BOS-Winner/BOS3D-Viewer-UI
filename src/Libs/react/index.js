import React from "react";
import useUpdateEffect from "./useUpdateEffect";

React.useUpdateEffect = useUpdateEffect;

export default React;
// export { default as useUpdateEffect } from "./useUpdateEffect";
