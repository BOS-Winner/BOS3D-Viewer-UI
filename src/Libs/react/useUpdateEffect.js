import { useEffect, useRef } from "react";

export default function (effect, deps) {
  const isFirstRender = useRef(true);
  useEffect(() => {
    if (isFirstRender.current) {
      isFirstRender.current = false;
    } else {
      effect();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, deps);
}
