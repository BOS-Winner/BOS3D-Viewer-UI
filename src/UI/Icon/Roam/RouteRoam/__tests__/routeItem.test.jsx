import React from 'react';
import { render, act } from "@testing-library/react";
import BOS3D from 'BOS3D';
import RouteItem from '../RouteItem';

const viewer = new BOS3D.Viewer();
const onDelete = jest.fn();
const onRename = jest.fn();
const onPlay = jest.fn();
const onChangeSpeed = jest.fn();
const onChangeTime = jest.fn();
const onExport = jest.fn();
const roamPlayer = {
  addStopPlayCallback: () => { },
  addKeyFrameCallback: () => { },
};

let riContainer;
let riGetByTitle;
let riGetByPlaceholderText;
let riGetByText;
let riRerender;

describe('route item', () => {
  beforeEach(() => {
    const {
      container, getByTitle, getByPlaceholderText, getByText, rerender
    } = render((
      <RouteItem
        name="test"
        viewer={viewer}
        onDelete={onDelete}
        onRename={onRename}
        onPlay={onPlay}
        onChangeTime={onChangeTime}
        onChangeSpeed={onChangeSpeed}
        onChangeFrame={() => {}}
        onExport={onExport}
        roamPlayer={roamPlayer}
        playingState="stop"
        recTimeLen={10}
        routeLen={11}
      />
    ));

    riContainer = container;
    riGetByTitle = getByTitle;
    riGetByPlaceholderText = getByPlaceholderText;
    riGetByText = getByText;
    riRerender = rerender;
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should rename item', () => {
    act(() => {
      riGetByTitle('重命名').click();
    });
    expect(riContainer).toMatchSnapshot();

    act(() => {
      riGetByPlaceholderText('test').value = 'rename';
      riGetByText('确定').click();
    });
    expect(onRename).toBeCalledTimes(1);
    expect(onRename).lastCalledWith('rename');

    act(() => {
      riGetByTitle('重命名').click();
    });
    act(() => {
      riGetByText('取消').click();
    });
    expect(onRename).toBeCalledTimes(1);

    riRerender((
      <RouteItem
        name="rename"
        viewer={viewer}
        onDelete={onDelete}
        onRename={onRename}
        onPlay={onPlay}
        onChangeTime={onChangeTime}
        onChangeSpeed={onChangeSpeed}
        onExport={onExport}
        roamPlayer={roamPlayer}
        onChangeFrame={() => {}}
        playingState="stop"
        recTimeLen={10}
        routeLen={11}
      />
    ));
    expect(riGetByText('rename')).not.toBeUndefined();
  });

  it('should call functions', () => {
    riGetByTitle('删除').click();
    expect(onDelete).toBeCalledTimes(1);
    riGetByTitle('导出').click();
    expect(onExport).toBeCalledTimes(1);
    riGetByTitle('播放').click();
    expect(onPlay).toBeCalledTimes(1);
    expect(onPlay).lastCalledWith('play');
    riRerender((
      <RouteItem
        name="test"
        viewer={viewer}
        onDelete={onDelete}
        onRename={onRename}
        onPlay={onPlay}
        onChangeTime={onChangeTime}
        onChangeSpeed={onChangeSpeed}
        onExport={onExport}
        onChangeFrame={() => {}}
        roamPlayer={roamPlayer}
        playingState="play"
        recTimeLen={10}
        routeLen={11}
      />
    ));
    riGetByTitle('暂停').click();
    expect(onPlay).toBeCalledTimes(2);
    expect(onPlay).lastCalledWith('pause');
  });

  it('should change speed', () => {
    for (let index = 0; index < 5; index++) {
      // eslint-disable-next-line no-loop-func
      act(() => {
        riGetByText('+').click();
      });
    }
    expect(riGetByText('1.5')).not.toBeUndefined();
    for (let index = 0; index < 4; index++) {
      // eslint-disable-next-line no-loop-func
      act(() => {
        riGetByText('-').click();
      });
    }
    expect(riGetByText('1.1')).not.toBeUndefined();
  });
  // TODO: test ranger
});
