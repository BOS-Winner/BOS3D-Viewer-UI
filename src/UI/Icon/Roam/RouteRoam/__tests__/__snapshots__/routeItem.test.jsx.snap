// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`route item should rename item 1`] = `
<div>
  <div
    class="routeItem"
  >
    <div
      class="row1"
    >
      <div
        class="name"
        role="presentation"
      >
        <input
          maxlength="50"
          placeholder="test"
          type="text"
        />
        <button
          type="button"
        >
          确定
        </button>
        <button
          type="button"
        >
          取消
        </button>
      </div>
      <div>
        <img
          alt="播放"
          src="play.png"
          title="播放"
        />
        <img
          alt="删除"
          src="remove.png"
          title="删除"
        />
        <img
          alt="导出"
          src="export.png"
          title="导出"
        />
      </div>
    </div>
    <div
      class="row2"
    >
      <div
        class="container"
      >
        <div
          class="minButton "
          role="button"
          tabindex="0"
        >
          -
        </div>
        <div
          class="number"
        >
          1
        </div>
        <div
          class="maxButton "
          role="button"
          tabindex="0"
        >
          +
        </div>
      </div>
      <span>
        00:00:00
      </span>
      <div
        class="container"
        role="presentation"
      >
        <div
          class="rail"
        />
        <div
          class="track"
          style="width: 0%;"
        />
        <div
          class="step"
          style="left: 0%;"
        />
      </div>
    </div>
  </div>
</div>
`;
